### ISSUE TITLE
Model with a root validator in `Depends` leads to 500 error instead of 422

### ISSUE THREAD

**Issue by bsamseth** [tiangolo/fastapi#4974]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

from fastapi import FastAPI
from fastapi.param_functions import Depends
from pydantic import BaseModel
from pydantic.class_validators import root_validator

app = FastAPI()


class Options(BaseModel):
    a: Optional[int]
    b: Optional[int]

    @root_validator()
    def at_least_one_option(cls, values):
        assert any(
            opt is not None for opt in values.values()
        ), "At least one option should be enabled."
        return values


@app.get("/")
def read_root(options: Options = Depends(Options)):
    return options
```

### Description

If the assertion in the `root_validator` fails, the validation error that is produced is not caught correctly. The result is a 500 error to the client, and a traceback showing the validation error on the server-side. The expected result would be a 422 response, similar to how any other validation error is handled.

Reproduce as follows:

* Save the above snippet to a file, say `main.py`
* Run `uvicorn main:app` or some other way of running the app.
* In a separate terminal, run `curl -X 'GET' '127.0.0.1:8000/?a=str' -H 'accept: application/json'`
  * This produces a 422 response as expected.
* Now run `curl -X 'GET' '127.0.0.1:8000/' -H 'accept: application/json'`
  * This produces a 500 error
  * Meanwhile, the other terminal will show an error traceback ending in a validation error.

I believe this is caused by the error being triggered during dependency resolution, which appears to be outside the scope of the error handler that catches validation errors.

### Operating System

Linux

### FastAPI Version

0.78.0

### Python Version

3.9.1

### Additional Context

I'm doing this because it allows defining query parameters with a pydantic model, which allows me to reuse this model as part of other models. Based on https://fastapi.tiangolo.com/tutorial/dependencies/classes-as-dependencies/

**Comment by Lin-Your** [comment]
Try this.
```python
from typing import Optional

from fastapi import FastAPI, HTTPException
from fastapi.param_functions import Depends
from pydantic import BaseModel
from pydantic.class_validators import root_validator

app = FastAPI()


class Options(BaseModel):
    a: Optional[int]
    b: Optional[int]

    @root_validator()
    def at_least_one_option(cls, values):
        if all(opt is None for opt in values.values()):
            raise HTTPException(422, "At least one option should be enabled.")
        # assert any(
        #     opt is not None for opt in values.values()
        # ), "At least one option should be enabled."
        return values


@app.get("/")
def read_root(options: Options = Depends(Options)):
    return options
```

**Comment by Lin-Your** [comment]
Or, if you want to catch all `AssertionError` and return a 422 response, it might help you.
```python
from typing import Optional

from fastapi import Depends, FastAPI, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel

app = FastAPI()


@app.exception_handler(AssertionError)
async def assertion_exception_handler(request: Request, exc: AssertionError):
    return JSONResponse(status_code=422, content={'detail': str(exc)})


class Options(BaseModel):
    a: Optional[int]
    b: Optional[int]


@app.get("/")
def read_root(options: Options = Depends(Options)):
    assert any(opt is not None for opt in options.dict().values()), "At least one option should be enabled."
    return options
```

**Comment by bsamseth** [comment]
@Lin-Your Thank's for the suggestions. To work around this I instead added a normal pydantic validator to the last field in the model, and did the assertion there. That way I can still rely on the built-in error handling and formatting of validation errors.

I didn't make this issue because I need a workaround, however, and rather because this looks like a straight-up bug. While one can certainly do other things to achieve similar results, I still think this should be fixed.

**Comment by JarroVGIT** [comment]
I agree, it is a straight up bug. I played around a bit and it seems that somehow Pydantic is throwing an `ValidationError` that isn't handled by Fastapi. Normally, any `ValidationError` that is thrown by Pydantic is caught and raised as an `RequestValidationError` by Fastapi, which is then by default handled by `request_validation_exception_handler` (returning a 422 response). I don't see how to solve this though. 

To satisfy my curiosity; you said:

> I instead added a normal pydantic validator to the last field in the model, and did the assertion there. 

Could you show me how? I tried to reproduce this but I got the same behaviour with `@validator`. Only when I did something like `a : Optional[int] = Field(..., gt=10)` I was able to get a 'real' 422 (when providing an a<10).

**Comment by JarroVGIT** [comment]
Just wanted to let you know that this is a known issue, and a long standing feature request. See this issue for workarounds and details: #1474
